docs/design/async-checkpoint.md

```diff
@@ -120,6 +124,93 @@ training process to resume the training, the training process can load the check
 than the storage.
 
+## The ElasticAgent Asynchronously Saves the Checkpoint into Storage
+
+If we start a daemon subprocess of the GPU training process, the daemon
+subprocess will exit if the training process fails. In this case, the
+parameters in CPU memory of the daemon subprocess will be cleaned. So, the elastic agent
+in the main process can allocate the shared memory with the training
+process. The training process saves the state dict to the shared memory and
+the agent save them into the storage.
+
+### The Classes Design
+
+As we need a global thread to keep and sync the checkpointing state in storage, and an agent
+thread per node to save the checkpointing state into the storage, we design three classes to
+implement the checkpointing process.
+
+<div align="center">
+<img src="../figures/async-ckpt-classes.jpg" alt="Async Checkpoint Classes" width="1000">
+</div>
+
+- **GlobalCkptManager**
+    - Only one instance runs in the Rank 0 Agent process.
+    - The class is responsible for keeping the consistency of the checkpointing state in the shared
+      memory and the storage.
+    - Get the read lock of shared memory and notify all agents to save the checkpointing state
+      into the storage, update the checkpointing tracker file in the storage.
+- **AgentCkptManger**
+    - One instance runs in each agent process.
+    - Responsible for saving the checkpointing state from shared memory into the storage.
+    - Notify the GlobalCkptManager when the saving procedure is completed.
+- **TrainCkptManger**
+    - One instance runs in each training process.
+    - Responsible for saving the checkpointing state from GPU to shared memory.
+    - Notify the GlobalCkptManager when we need to save the checkpointing state into the storage.
+
+### Async Checkpointing Saving Steps
+
+<div align="center">
+<img src="../figures/async-ckpt-steps.jpg" alt="Async Checkpoint Classes" width="1000">
+</div>
+
+The agent and training process need to do the following steps:
+
+- Init Process - first checkpoint to memory
+
+    - The agent monitors the training process to create the mata of model and
+      optimizer state dict.
+    - The training process notifies GlobalCkptManager the checkpointing meta when the
+      training process firstly saves the state dict into the memory.
+    - GlobalCkptManager acquire the lock of the shared memory and notify all agents to
+      allocate the shared memory with the checkpointing meta.
+    - The TrainCkptManger acquires the lock of the shared memory and copies the state dict
+      from GPU to the shared memory.
+
+- Training Loop
+
+    - TrainCkptManager acquires the lock of the shared memory and copies the state dict
+      from GPU to the shared memory.
+    - TrainCkptManager notifies GlobalCkptManager to save the checkpointing state into the
+      storage if the iteration step is multiple of the save_interval.
+    - GlobalCkptManager acquires the lock of the shared memory and notifies all agents to
```

Comment:
It is not necessary to notify all agents. The agent saves its state dict to the storage when the training process notify it to save.

Reply:
for simplicity and stability, we use a polling instead of notification.